File: parte4_frontend/web/src/demo/privacidad.ts
/**
 * Filtro de privacidad de la API de acceso (`parte2_plataforma/comun/privacidad.py`), portado para el modo
 * demostración: el portal publicado en Vercel no tiene API detrás y este módulo decide igual que ella qué consulta
 * se responde, cuál se rechaza y qué alternativa se propone. Las reglas se leen de `config/privacidad.json`, el mismo
 * fichero que usan la API y Spark, y `privacidad.test.ts` compara las decisiones con respuestas reales de la API.
 *
 * Solo decide y da forma: los datos que acaban en la respuesta son los que la API ya devolvió enmascarados.
 * Las fechas son ISO sin zona (`2020-01-15T08:00:00`) y se operan como milisegundos UTC.
 */
import reglas from '../../../../config/privacidad.json'

import type { Nivel } from '@/api/tipos'

export const REGLAS = reglas
export const K_MINIMO = reglas.k_minimo
export const MAX_FILAS = reglas.max_filas_por_respuesta
export const NOTA =
  'Los grupos enmascarados no se suman a ningún total: ocultarlos no serviría si se pudieran deducir por diferencia.'

/** Los de `privacidad.py`: la API no los lee de la configuración. */
export const BARRIOS = ['Manhattan', 'Brooklyn', 'Queens', 'Bronx', 'Staten Island', 'EWR', 'N/A', 'Unknown']
const NIVELES = Object.keys(reglas.niveles) as Nivel[]
const FUENTES = Object.keys(reglas.fuentes)

const HORA = 3_600_000
const DIA = 24 * HORA

/** La `Consulta` de la API ya validada, con sus valores por defecto (así la devuelve la API en `Respuesta.consulta`). */
export interface ConsultaApi {
  nivel: Nivel
  fuente: 'historico' | 'tiempo_real'
  desde: string
  hasta: string
  metricas: string[]
  zona_origen: number | null
  barrio_origen: string | null
  barrio_destino: string | null
  campos_extra: string[]
}

/** Un error de validación con la forma de los de FastAPI (`{"detail": [ErrorValidacion, …]}`). */
export interface ErrorValidacion {
  type: string
  loc: (string | number)[]
  msg: string
  input: unknown
}

export type Decision =
  | { resultado: 'permitida'; motivos: string[]; alternativa: null }
  | { resultado: 'rechazada'; motivos: string[]; alternativa: ConsultaApi }

// --- fechas ------------------------------------------------------------------------------------------------

const FECHA = /^(\d{4})-(\d{2})-(\d{2})(?:[T ](\d{2}):(\d{2})(?::(\d{2}))?)?$/

/** Milisegundos UTC de una fecha ISO sin zona, o null si no lo es. */
export function aMilisegundos(texto: unknown): number | null {
  const partes = typeof texto === 'string' ? FECHA.exec(texto.trim()) : null
  if (!partes) return null
  const [anio, mes, dia, hora = '0', minuto = '0', segundo = '0'] = partes.slice(1)
  const ms = Date.UTC(+anio, +mes - 1, +dia, +hora, +minuto, +segundo)
  const fecha = new Date(ms)
  // 2020-02-31 no es una fecha: Date.UTC la convertiría en el 2 de marzo
  if (fecha.getUTCMonth() !== +mes - 1 || fecha.getUTCDate() !== +dia || +hora > 23 || +minuto > 59 || +segundo > 59) {
    return null
  }
  return ms
}

/** `2020-01-15T08:00:00` */
export function aIso(ms: number): string {
  return new Date(ms).toISOString().slice(0, 19)
}

// --- validación (el modelo pydantic `Consulta`) --------------------------------------------------------------

function listaTolerante(valor: unknown): string[] {
  let lista: unknown[]
  if (valor === null || valor === undefined || valor === '') {
    lista = []
  } else if (typeof valor === 'string') {
    try {
      const cargado: unknown = JSON.parse(valor)
      lista = Array.isArray(cargado) ? cargado : [cargado]
    } catch {
      lista = valor.trim().replace(/^[[\]]+|[[\]]+$/g, '').split(',')
    }
  } else {
    lista = Array.isArray(valor) ? valor : [valor]
  }
  return lista.map((v) => String(v).trim().replace(/^['"]+|['"]+$/g, '')).filter(Boolean)
}

function barrio(valor: unknown): string | null {
  if (valor === null || valor === undefined) return null
  return String(valor).trim() || null
}

/** Valida el cuerpo como lo hace FastAPI con `Consulta`: la consulta normalizada o la lista de errores (422). */
export function validar(cuerpo: unknown): { consulta: ConsultaApi } | { errores: ErrorValidacion[] } {
  if (!cuerpo || typeof cuerpo !== 'object' || Array.isArray(cuerpo)) {
    return { errores: [{ type: 'model_attributes_type', loc: ['body'], msg: 'Input should be a valid dictionary', input: cuerpo }] }
  }
  const datos = cuerpo as Record<string, unknown>
  const errores: ErrorValidacion[] = []
  const error = (campo: string, type: string, msg: string) =>
    errores.push({ type, loc: ['body', campo], msg, input: datos[campo] })

  if (datos.nivel === undefined) error('nivel', 'missing', 'Field required')
  else if (!NIVELES.includes(datos.nivel as Nivel)) {
    error('nivel', 'literal_error', "Input should be 'hora_zona', 'dia_barrio' or 'od_dia_barrio'")
  }
  const fuente = datos.fuente ?? 'historico'
  if (!FUENTES.includes(fuente as string)) error('fuente', 'literal_error', "Input should be 'historico' or 'tiempo_real'")

  const fechas: Record<'desde' | 'hasta', number | null> = { desde: null, hasta: null }
  for (const campo of ['desde', 'hasta'] as const) {
    if (datos[campo] === undefined) error(campo, 'missing', 'Field required')
    else if ((fechas[campo] = aMilisegundos(datos[campo])) === null) {
      error(campo, 'datetime_from_date_parsing', 'Input should be a valid datetime or date')
    }
  }

  let zona: number | null = null
  const zonaPedida = datos.zona_origen
  if (!(zonaPedida === null || zonaPedida === undefined || zonaPedida === '' || zonaPedida === 'null' || zonaPedida === 'None')) {
    const numero = typeof zonaPedida === 'boolean' ? Number(zonaPedida) : Number(String(zonaPedida).trim())
    if (typeof zonaPedida === 'string' && !/^\s*[+-]?\d+\s*$/.test(zonaPedida)) {
      error('zona_origen', 'int_parsing', 'Input should be a valid integer, unable to parse string as an integer')
    } else if (!Number.isInteger(numero)) {
      error('zona_origen', 'int_from_float', 'Input should be a valid integer, got a number with a fractional part')
    } else {
      zona = numero
    }
  }

  if (errores.length) return { errores }
  const metricas = listaTolerante(datos.metricas)
  return {
    consulta: {
      nivel: datos.nivel as Nivel,
      fuente: fuente as ConsultaApi['fuente'],
      desde: aIso(fechas.desde!),
      hasta: aIso(fechas.hasta!),
      metricas: metricas.length ? metricas : ['n_viajes'],
      zona_origen: zona,
      barrio_origen: barrio(datos.barrio_origen),
      barrio_destino: barrio(datos.barrio_destino),
      campos_extra: listaTolerante(datos.campos_extra),
    },
  }
}

// --- decisión (`privacidad.evaluar`) ---------------------------------------------------------------------

function paso(nivel: Nivel): number {
  return reglas.niveles[nivel].tiempo === 'hora' ? HORA : DIA
}

function alinear(momento: number, intervalo: number, arriba: boolean): number {
  let base = Math.floor(momento / HORA) * HORA
  if (intervalo >= DIA) base = Math.floor(momento / DIA) * DIA
  if (arriba && base < momento) base += intervalo
  return base
}

/** `repr()` de Python para un texto, como en los motivos de la API. */
function repr(texto: string): string {
  if (texto.includes("'") && !texto.includes('"')) return `"${texto}"`
  return `'${texto.replace(/\\/g, '\\\\').replace(/'/g, "\\'")}'`
}

/** Decide si la consulta se puede responder; `enmascarada` se decide después, al ver los datos. */
export function evaluar(consulta: ConsultaApi): Decision {
  const motivos: string[] = []
  const alt: ConsultaApi = { ...consulta, metricas: [...consulta.metricas], campos_extra: [] }
  const individuales = new Set(reglas.campos_individuales)

  const pedidosIndividuales = [...new Set(consulta.campos_extra.filter((c) => individuales.has(c)))].sort()
  if (pedidosIndividuales.length) motivos.push(`pide campos individuales: ${pedidosIndividuales.join(', ')}`)
  const desconocidos = [...new Set(consulta.campos_extra.filter((c) => !individuales.has(c)))].sort()
  if (desconocidos.length) motivos.push(`campos no publicados: ${desconocidos.join(', ')}`)

  const noPublicadas = consulta.metricas.filter((m) => !reglas.metricas.includes(m))
  if (noPublicadas.length) {
    motivos.push(`métricas no publicadas: ${noPublicadas.join(', ')}`)
    const publicadas = consulta.metricas.filter((m) => reglas.metricas.includes(m))
    alt.metricas = publicadas.length ? publicadas : ['n_viajes']
  }

  // primero el nivel al que hay que subir; después, los filtros que ese nivel no admite
  const dimensiones = reglas.niveles[consulta.nivel].dimensiones
  if (consulta.barrio_destino !== null && !dimensiones.includes('barrio_destino')) {
    motivos.push('el destino solo se publica por barrio y día (nivel od_dia_barrio)')
    alt.nivel = 'od_dia_barrio'
  } else if (consulta.barrio_origen !== null && !dimensiones.includes('barrio_origen')) {
    motivos.push(`el nivel ${consulta.nivel} no filtra por barrio; se usa dia_barrio`)
    alt.nivel = 'dia_barrio'
  }
  for (const campo of ['barrio_origen', 'barrio_destino'] as const) {
    const valor = consulta[campo]
    if (valor !== null && !BARRIOS.includes(valor)) {
      motivos.push(`${campo} desconocido: ${repr(valor)} (los barrios son: ${BARRIOS.slice(0, 6).join(', ')})`)
      alt[campo] = null
    }
  }

  if (consulta.zona_origen !== null && !reglas.niveles[alt.nivel].dimensiones.includes('zona_origen')) {
    motivos.push(`el nivel ${alt.nivel} no tiene zona; se agrega por barrio`)
    alt.zona_origen = null
  }

  const intervalo = paso(alt.nivel)
  const pedidoDesde = aMilisegundos(consulta.desde)!
  const pedidoHasta = aMilisegundos(consulta.hasta)!
  let altHasta = pedidoHasta
  if (pedidoHasta <= pedidoDesde) {
    motivos.push('la ventana temporal está vacía')
    altHasta = pedidoDesde + intervalo
  }
  const desde = alinear(pedidoDesde, intervalo, false)
  let hasta = alinear(altHasta, intervalo, true)
  if (hasta - desde < intervalo) hasta = desde + intervalo
  if (desde !== pedidoDesde || hasta !== pedidoHasta) {
    motivos.push(`la granularidad mínima es de ${intervalo === HORA ? 'una hora completa' : 'un día completo'}`)
  }
  const maximo = reglas.max_dias_por_consulta * DIA
  if (hasta - desde > maximo) {
    motivos.push(`el rango máximo por consulta es de ${reglas.max_dias_por_consulta} días`)
    hasta = desde + maximo
  }
  alt.desde = aIso(desde)
  alt.hasta = aIso(hasta)

  if (motivos.length) return { resultado: 'rechazada', motivos, alternativa: alt }
  return { resultado: 'permitida', motivos: [], alternativa: null }
}

// --- respuesta (`repositorio.proyeccion` y `privacidad.enmascarar`) ---------------------------------------------

export type FilaGuardada = Record<string, unknown>

/** Campo temporal del nivel (`hora` o `dia`). */
export function campoTiempo(nivel: Nivel): 'hora' | 'dia' {
  return nivel === 'hora_zona' ? 'hora' : 'dia'
}

/** ¿La fila cumple los filtros de la consulta (ventana y dimensiones)? Como `repositorio.filtro_mongo`. */
export function cumple(fila: FilaGuardada, consulta: ConsultaApi): boolean {
  const tiempo = String(fila[campoTiempo(consulta.nivel)])
  if (tiempo < consulta.desde || tiempo >= consulta.hasta) return false
  return (['zona_origen', 'barrio_origen', 'barrio_destino'] as const).every(
    (campo) => consulta[campo] === null || fila[campo] === consulta[campo],
  )
}

/** Solo los campos que la API proyecta para esta consulta. */
export function proyectar(fila: FilaGuardada, consulta: ConsultaApi): FilaGuardada {
  const extra = consulta.nivel === 'hora_zona' ? ['zona_origen_nombre', 'barrio_origen'] : []
  const campos = new Set([...reglas.niveles[consulta.nivel].dimensiones, ...extra, ...consulta.metricas, 'n_viajes', 'suprimido'])
  return Object.fromEntries(Object.entries(fila).filter(([campo, valor]) => campos.has(campo) && valor !== undefined))
}

/** Oculta las cifras de los grupos suprimidos: `n_viajes` pasa a `"oculto"` y el resto de métricas a null. */
export function enmascarar(filas: FilaGuardada[], metricas: string[]): { filas: FilaGuardada[]; ocultas: number } {
  let ocultas = 0
  const salida = filas.map((fila) => {
    if (!fila.suprimido) return fila
    ocultas += 1
    const oculta: FilaGuardada = { ...fila }
    for (const metrica of metricas) oculta[metrica] = null
    oculta.n_viajes = 'oculto'
    return oculta
  })
  return { filas: salida, ocultas }
}

File: parte2_plataforma/comun/privacidad.py
"""Reglas propias de privacidad (E3) para las consultas: qué se permite, qué se enmascara y qué se rechaza.

La protección tiene dos capas:
  1. Spark solo publica agregados y deja marcados como `suprimido` los grupos con menos de
     `k_minimo` viajes (sin cifras). Los datos individuales no salen nunca de la zona restringida.
  2. Este módulo filtra cada consulta antes de ir a MongoDB: solo niveles de agregación
     publicados, ventanas alineadas a su granularidad, rango máximo, y ningún campo individual.
     Si rechaza, propone una alternativa agregada que sí se puede responder.

Todo es lógica pura (sin base de datos) para poder probarla a fondo. Las reglas están en
`config/privacidad.json`.
"""
from __future__ import annotations

import json
import os
from datetime import datetime, timedelta
from enum import StrEnum
from functools import lru_cache
from pathlib import Path
from typing import Literal

from pydantic import BaseModel, Field, field_validator

RUTA_CONFIG = Path(os.environ.get('PIDS_CONFIG_DIR', Path(__file__).resolve().parents[2] / 'config'))


@lru_cache(maxsize=1)
def config() -> dict:
    return json.loads((RUTA_CONFIG / 'privacidad.json').read_text(encoding='utf-8'))


class Resultado(StrEnum):
    PERMITIDA = 'permitida'
    ENMASCARADA = 'enmascarada'
    RECHAZADA = 'rechazada'


BARRIOS = ['Manhattan', 'Brooklyn', 'Queens', 'Bronx', 'Staten Island', 'EWR', 'N/A', 'Unknown']


class Consulta(BaseModel):
    """Consulta agregada. Es lo único que acepta la API de acceso."""
    nivel: Literal['hora_zona', 'dia_barrio', 'od_dia_barrio']
    fuente: Literal['historico', 'tiempo_real'] = 'historico'
    desde: datetime
    hasta: datetime
    metricas: list[str] = Field(default_factory=lambda: ['n_viajes'])
    zona_origen: int | None = None
    barrio_origen: str | None = None
    barrio_destino: str | None = None
    campos_extra: list[str] = Field(default_factory=list,
                                    description='Campos adicionales pedidos; si alguno es individual, se rechaza')

    @field_validator('barrio_origen', 'barrio_destino', mode='before')
    @classmethod
    def _limpiar_barrio(cls, valor):
        """Un LLM manda a veces "" o una lista en texto: se trata como "sin filtro"."""
        if valor is None:
            return None
        texto = str(valor).strip()
        return texto or None

    @field_validator('zona_origen', mode='before')
    @classmethod
    def _limpiar_zona(cls, valor):
        return None if valor in (None, '', 'null', 'None') else valor

    @field_validator('metricas', 'campos_extra', mode='before')
    @classmethod
    def _lista_tolerante(cls, valor, info):
        """Un LLM manda a veces la lista como texto ('["n_viajes"]' o 'n_viajes, propina_media').

        Se acepta la forma chapucera y se normaliza: ser tolerante con el formato no relaja ninguna
        regla de privacidad, porque después se valida contra las métricas publicadas.
        """
        if valor is None or valor == '':
            valor = []
        elif isinstance(valor, str):
            try:
                cargado = json.loads(valor)
            except ValueError:
                cargado = valor.strip().strip('[]').split(',')
            valor = cargado if isinstance(cargado, list) else [cargado]
        elif not isinstance(valor, (list, tuple, set)):
            valor = [valor]
        limpio = [str(v).strip().strip('\'"') for v in valor]
        limpio = [v for v in limpio if v]
        if info.field_name == 'metricas' and not limpio:
            return ['n_viajes']
        return limpio


class Decision(BaseModel):
    resultado: Resultado
    motivos: list[str] = Field(default_factory=list)
    alternativa: Consulta | None = None

    @property
    def permitida(self) -> bool:
        return self.resultado != Resultado.RECHAZADA


def coleccion(consulta: Consulta) -> str:
    cfg = config()
    return cfg['fuentes'][consulta.fuente] + cfg['niveles'][consulta.nivel]['coleccion']


def _paso(nivel: str) -> timedelta:
    return timedelta(hours=1) if config()['niveles'][nivel]['tiempo'] == 'hora' else timedelta(days=1)


def _alinear(momento: datetime, paso: timedelta, arriba: bool) -> datetime:
    base = momento.replace(minute=0, second=0, microsecond=0)
    if paso >= timedelta(days=1):
        base = base.replace(hour=0)
    if arriba and base < momento:
        base += paso
    return base


def evaluar(consulta: Consulta) -> Decision:
    """Decide si la consulta se puede responder. `enmascarada` se decide después, al ver los datos."""
    cfg = config()
    motivos: list[str] = []
    alt = consulta.model_copy(deep=True)

    individuales = sorted(set(consulta.campos_extra) & set(cfg['campos_individuales']))
    if individuales:
        motivos.append(f'pide campos individuales: {", ".join(individuales)}')
    desconocidos = sorted(set(consulta.campos_extra) - set(cfg['campos_individuales']))
    if desconocidos:
        motivos.append(f'campos no publicados: {", ".join(desconocidos)}')
    alt.campos_extra = []

    no_publicadas = [m for m in consulta.metricas if m not in cfg['metricas']]
    if no_publicadas:
        motivos.append(f'métricas no publicadas: {", ".join(no_publicadas)}')
        alt.metricas = [m for m in consulta.metricas if m in cfg['metricas']] or ['n_viajes']

    # primero el nivel al que hay que subir; después, los filtros que ese nivel no admite
    dims = cfg['niveles'][consulta.nivel]['dimensiones']
    if consulta.barrio_destino is not None and 'barrio_destino' not in dims:
        motivos.append('el destino solo se publica por barrio y día (nivel od_dia_barrio)')
        alt.nivel = 'od_dia_barrio'
    elif consulta.barrio_origen is not None and 'barrio_origen' not in dims:
        motivos.append(f'el nivel {consulta.nivel} no filtra por barrio; se usa dia_barrio')
        alt.nivel = 'dia_barrio'
    for campo in ('barrio_origen', 'barrio_destino'):
        valor = getattr(consulta, campo)
        if valor is not None and valor not in BARRIOS:
            barrios = ', '.join(BARRIOS[:6])
            motivos.append(f'{campo} desconocido: {valor!r} (los barrios son: {barrios})')
            setattr(alt, campo, None)

    dims_alt = cfg['niveles'][alt.nivel]['dimensiones']
    if consulta.zona_origen is not None and 'zona_origen' not in dims_alt:
        motivos.append(f'el nivel {alt.nivel} no tiene zona; se agrega por barrio')
        alt.zona_origen = None

    paso = _paso(alt.nivel)
    if consulta.hasta <= consulta.desde:
        motivos.append('la ventana temporal está vacía')
        alt.hasta = alt.desde + paso
    desde = _alinear(alt.desde, paso, arriba=False)
    hasta = _alinear(alt.hasta, paso, arriba=True)
    if hasta - desde < paso:
        hasta = desde + paso
    if (desde, hasta) != (consulta.desde, consulta.hasta):
        unidad = 'una hora completa' if paso == timedelta(hours=1) else 'un día completo'
        motivos.append(f'la granularidad mínima es de {unidad}')
    maximo = timedelta(days=cfg['max_dias_por_consulta'])
    if hasta - desde > maximo:
        motivos.append(f'el rango máximo por consulta es de {cfg["max_dias_por_consulta"]} días')
        hasta = desde + maximo
    alt.desde, alt.hasta = desde, hasta

    if motivos:
        return Decision(resultado=Resultado.RECHAZADA, motivos=motivos, alternativa=alt)
    return Decision(resultado=Resultado.PERMITIDA)


def rechazo_individual(descripcion: str) -> Decision:
    """Respuesta fija para cualquier intento de obtener viajes concretos."""
    return Decision(
        resultado=Resultado.RECHAZADA,
        motivos=[f'petición de datos individuales: {descripcion}'[:300],
                 'la plataforma solo publica agregados de al menos '
                 f'{config()["k_minimo"]} viajes'],
    )


def enmascarar(filas: list[dict], metricas: list[str]) -> tuple[list[dict], int]:
    """Oculta las cifras de los grupos suprimidos. Devuelve las filas y cuántas se han enmascarado.

    La etiqueta es ``oculto`` para todos: un grupo suprimido puede tener menos de k viajes o ser
    el complementario (10 o más), y esa marca no se publica, así que ``<10`` sería falso. Nunca se
    suman a ningún total: el total menos lo visible revelaría justo lo oculto (ataque por diferencia).
    """
    salida, ocultas = [], 0
    for fila in filas:
        fila = {c: v for c, v in fila.items() if c != '_id'}
        if fila.get('suprimido'):
            ocultas += 1
            for m in metricas:
                fila[m] = None
            fila['n_viajes'] = 'oculto'
        salida.append(fila)
    return salida, ocultas


def resultado_final(ocultas: int) -> Resultado:
    return Resultado.ENMASCARADA if ocultas else Resultado.PERMITIDA
